homework/ba-nguyen/lesson-8/javascript-4.md

```diff
@@ -0,0 +1,273 @@
+Use the === operator, or something equivalent such as indexOf, to determine whether two values are the same.
+Give the class a static from method that takes an iterable object as argument and creates a group that contains all the values produced by iterating over it.
+
+```ts
+class Group<T> {
+  private array: T[];
+  constructor(array: T[]) {
+    this.array = [];
+
+    for (const numb of array) {
+      this.add(numb);
+    }
+  }
+
+  static from<T>(array: T[]): Group<T> {
+    return new Group(array);
+  }
+
+  has(numb: T): boolean {
+    return this.array.indexOf(numb) > -1;
+  }
+
+  add(numb: T): void {
+    if (this.has(numb)) return;
+    this.array.push(numb);
+  }
+
+  delete(numb: T): void {
+    if (this.has(numb)) {
+      const index = this.array.indexOf(numb);
+      this.array.splice(index, 1);
+    }
+  }
+}
+
+let group = Group.from([10, 20]);
+console.log(group.has(10));
+// → true
+console.log(group.has(30));
+// → false
+group.add(10);
+group.delete(10);
+console.log(group.has(10));
+// → false
+```
+
+**Question 2:** Write an expression that matches only JavaScript-style numbers. It must support an optional minus or plus sign in front of the number, the decimal dot, and exponent notation—5e-3 or 1E10—again with an optional sign in front of the exponent. Also note that it is not necessary for there to be digits in front of or after the dot, but the number cannot be a dot alone. That is, .5 and 5. are valid JavaScript numbers, but a lone dot isn’t.
+
+```js
+// Fill in this regular expression.
+let number =
+  /^(\+|-)*[\d]+((e|E)(\+|-)*[\d]+)*(\.(\+|-)*[\d]+(e|E)*((\+|-)*[\d]+)*){0,1}$/;
+
+// Tests:
+for (let str of [
+  "1",
+  "-1",
+  "+15",
+  "1.55",
+  ".5",
+  "5.",
+  "1.3e2",
+  "1E-4",
+  "1e+12",
+]) {
+  if (!number.test(str)) {
+    console.log(`Failed to match '${str}'`);
+  }
+}
+
+for (let str of ["1a", "+-1", "1.2.3", "1+1", "1e4.5", ".5.", "1f5", "."]) {
+  if (number.test(str)) {
+    console.log(`Incorrectly accepted '${str}'`);
+  }
+}
+```
+
+**Question 3**: Given an array of promises, Promise.all returns **_a_** promise that waits for all of the promises in the array to finish. It then succeeds, yielding an array of result values. If a promise in the array fails, the promise returned by all fails too, with the failure reason from the failing promise.
+Implement something like this yourself as a regular function called **_Promise_all_**.
+Remember that after a promise has succeeded or failed, it can’t succeed or fail again, and further calls to the functions that resolve it are ignored. This can simplify the way you handle failure of your promise.
+
+```js
+function Promise_all(promises) {
```

Comment:
```js
Promise_all([]).then(array => {
  console.log("This should be []:", array);
});
```
your output is difference  to expect output for this test case

Reply:
my output is right, `"This should be []:", []`, can you check again?

Reply:
@nguyenvanbapc66  can you give me a screenshot? I don't see console log

Reply:
![Result](https://user-images.githubusercontent.com/54839095/187017594-fa27378a-5997-4285-a9cb-20e11b4b6e68.png)

Reply:
@nguyenvanbapc66  Your code in the picture is different from the your code push =))

Reply:
Oh :)) sorry about that, i forgot to push the lastest code